### PR TITLE
Add Jax version of the original NumPy simulator

### PR DESCRIPTION
**Pull request by anigmetov**
Add JAX simulation path alongside the existing NumPy regression path.

  - JAXify Beam, BeamGauss, and SkyModels, and preserve the old implementations under lusee/pre_jax/ to test against them.
  - Add JaxSimulator class, keep the NumPy simulator for regression, and add NpWrapper for NumPy compatibility.
  - Switch beam harmonic evaluation to sphericart, which speeds up beam ALM generation.
  - Make CroSimulator stay on JAX arrays to ensure differentiability.
  - Add compatibility, smoke, 28-day regression for Jax simulator, and simulator gradient tests.

### CODE REVIEW

**Comment by xzackli** [comment]
@anigmetov could we take the sphericart addition out of this PR? I see it's segfaulting and would love to merge this ASAP. It's also not really related to the main PR functionality.

**Comment by xzackli** [comment]
I've made a small tweak to your excellent PR, so you can do this:

```python
sim = lusee.CroSimulator(obs, [beam], sky, Tground=0,
                         combinations=[(0,0)], freq=[10.0], lmax=8)

grad_sky = jax.grad(lambda s: jnp.sum(sim.simulate(sky=s) ** 2))(sky)
# grad_sky is a HarmonicPointSourceSky — same type, gradient values
grad_sky._T   # d(loss)/d(source amplitude)
```

**Comment by xzackli** [comment]
@anigmetov is this ready for review?

**Comment by anigmetov** [comment]
Now it is, I added the commits with your implementation. NB: I temporarily made it non-diff w.r.t. beam, to simplify the implementation.

File: lusee/pre_jax/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+"""Pre-JAX reference implementations preserved for migration testing."""
```
Comment (xzackli) [inline lusee/pre_jax/__init__.py:1]: Now that all tests pass (and presumably migration tests pass), I think these could be safely deleted.

File: lusee/pre_jax/BeamGauss.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import numpy as np
```
Comment (xzackli) [inline lusee/pre_jax/BeamGauss.py:1]: more pre_jax stuff that could be deleted if migration tests pass

File: lusee/pre_jax/MonoSkyModels.py [hunk 3]
```diff
@@ -0,0 +1,1035 @@
+import numpy as np
```
Comment (xzackli) [inline lusee/pre_jax/MonoSkyModels.py:1]: another option is to move these to tests/

File: lusee/Beam.py [hunk 4]
```diff
@@ -1,37 +1,141 @@
 #
 # LuSEE Beam
 #
+
+from functools import partial, lru_cache
+import os
+
+os.environ["JAX_ENABLE_X64"] = "True"
```
Comment (xzackli) [inline lusee/Beam.py]: We should not set the jax environmental variables

File: lusee/Beam.py [hunk 5]
```diff
@@ -228,49 +316,132 @@ class Beam:
     :param phi: Array of phi bins in radians
     :type phi: numpy array
     """
+    is_jax_pytree_beam = True
```
Comment (xzackli) [inline lusee/Beam.py:318]: we need to set BeamInterpolator to be one of these eventually

File: lusee/CroSimulator.py [hunk 6]
```diff
@@ -60,34 +60,94 @@ def __init__ (self, obs, beams, sky_model, Tground = 200.0,
...
+                groundPowerImag = -jnp.real(beamimag[:,0]) / jnp.sqrt(4*jnp.pi)
+            if "dump_beams" in getattr(self, "extra_opts", {}):
+                np.save(bi.id+bj.id, np.asarray(beamreal))
+            self.efbeams.append((i,j,beamreal, beamimag, groundPowerReal,
+                                 groundPowerImag))
 
             
                                 
-    def simulate(self, times=None):
+    def simulate(self, times=None, *, sky=None, beam=None):
```
Comment (xzackli) [inline lusee/CroSimulator.py:111]: This is some advanced pytree technology? Like tracing through kwarg?

File: lusee/frequencies.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import numpy as np
+import jax.numpy as jnp
+
+
+CANONICAL_FREQ_START_MHZ = 1.0
```
Comment (xzackli) [inline lusee/frequencies.py:5]: @slosar are these canonical frequencies correct

File: lusee/JaxSimulator.py [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+import sys
+import pickle
+import os
+import warnings
+from scipy.spatial.transform import Rotation as ScipyRotation
+import time
+from s2fft.recursions.risbo_jax import compute_full
+
+
+class JaxSimulator(SimulatorBase):
```
Comment (xzackli) [inline lusee/JaxSimulator.py]: I have a slightly unreasonable ask @anigmetov what if we rename this to WignerRotationSimulator or TopoSimulator or something? Jax doesn't define this anymore, the whole package has been ported.

File: lusee/NumpySimulator.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+import healpy as hp
+import fitsio
+import sys
+import pickle
+import os
+import warnings
+
+
+
+class DefaultSimulator(SimulatorBase):
```
Comment (xzackli) [inline lusee/NumpySimulator.py]: NumpySimulator is a great name

File: lusee/SkyModels.py [hunk 10]
```diff
@@ -149,15 +196,39 @@ class HealpixSky:
     :type frame: str
     
     """
-    def __init__ (self, Nside, lmax, maps, freq=[25.0], frame="galactic"):
+    def __init__ (self, Nside, lmax, maps, freq=None, frame="galactic"):
         self.Nside = Nside
         self.Npix = Nside**2 * 12
-        self.maps = maps
-        self.freq = freq
+        self.maps = jnp.asarray(maps)
```
Comment (xzackli) [inline lusee/SkyModels.py:198]: I'm wondering if we should actually discard this

File: lusee/SkyModels.py [hunk 11]
```diff
@@ -263,10 +346,10 @@ class HarmonicPointSourceSky:
     def __init__(self, lmax, freq, T=1.0, *,
                  ra_deg=None, dec_deg=None, l_deg=None, b_deg=None):
         self.lmax = lmax
-        self.freq = np.atleast_1d(freq)
-        T = np.atleast_1d(np.asarray(T, dtype=float))
+        self.freq = canonicalize_frequencies(freq, as_jax=True)
```
Comment (xzackli) [inline lusee/SkyModels.py:345]: what's the point of canonicalizing

File: simulation/driver/run_Cro_sim.py [hunk 12]
```diff
@@ -18,6 +19,21 @@ def __init__ (self,yaml):
         self._parse_sky()
         self._parse_beams()
 
+    @staticmethod
+    def _normalize_engine(cfg):
+        engine = cfg.get("simulation", {}).get("engine", "default")
+        e = str(engine).strip().lower()
+        aliases = {
```
Comment (xzackli) [inline simulation/driver/run_Cro_sim.py]: We should really have a discussion about the naming, this seems like a really good time to pick sane choices

File: simulation/driver/run_Cro_sim.py [hunk 13]
```diff
@@ -85,12 +106,14 @@ def _parse_beams(self):
             for b in self['observation']['beams']:
                 cbeam=bd[b]
                 print ("Creating gaussian beam",b,":")
-                B = lusee.BeamGauss(dec_deg=cbeam['declination'],
+                B = lusee.BeamGauss(alt_deg=cbeam['declination'],
                                       sigma_deg=cbeam['sigma'],
                                       one_over_freq_scaling=cbeam['one_over_freq_scaling'], id = b)
                 angle = bdc['common_beam_angle']+cbeam['angle']
                 print ("  rotating: ",angle)
                 B = B.rotate(angle)
```
Comment (xzackli) [inline simulation/driver/run_Cro_sim.py:111]: so B.rotate doesn't change it but some other B ops do? What's our mutation policy

File: simulation/driver/sim_driver.py [hunk 14]
```diff
@@ -6,11 +6,13 @@
...
 class SimDriver(dict):
     def __init__(self, cfg):
-        import lusee
-
-        self._lusee = lusee
         self.update(cfg)
+        # from simulator_ng
         self._resolve_simulation_paths()
+        # from jaxify
+        self._configure_jax_precision()
```
Comment (xzackli) [inline simulation/driver/sim_driver.py]: Should we really configure jax precision from the config? I am uncomfortable with different sources of truth

**Comment by xzackli** [comment]
One more review comment, did you test vs old simulator for (0,1) re/im antennas? @slosar arguably since we are overhauling that functionality we don't need it?

File: lusee/Beam.py (hunk 4 above)
Comment (slosar) [reply lusee/Beam.py]: I don't think we should be overriding them, but if not set probably ok to set defaults?

File: simulation/driver/run_Cro_sim.py (hunk 13 above)
Comment (slosar) [reply simulation/driver/run_Cro_sim.py:111]: I think we can make B totally immutable (i.e. B.rotate returns a new beam, which is the way to do this kind of shit). But I think this might also wait for the overhaul.

File: lusee/SkyModels.py (hunk 11 above)
Comment (anigmetov) [reply lusee/SkyModels.py:345]: The main point was to get rid of errors that I was getting in various places because the old code was looking up frequencies by value. Any minor numerical difference (32 bit vs 64 bit, linspace vs. arange) would lead to an error. So I decided it's better to have a single place where we define all 50 frequencies as a constant array; anywhere else we use indices (e.g., we want to look at each 5th frequency bin, you don't write `freqs = np.linspace(1.0, 50.0, 1.25)`, but `freq_inds = [ i for i in range(0, 50, 5)`). We can simplfiy this, if you want, removing all the functions and just subscripting the constant array.

File: lusee/JaxSimulator.py (hunk 8 above)
Comment (anigmetov) [reply lusee/JaxSimulator.py]: I am fine with any renaming.

File: lusee/Beam.py (hunk 4 above)
Comment (anigmetov) [reply lusee/Beam.py]: I am on the fence here. On the one hand, I would like to hard-code 64 bit precision everywhere. On the other hand, if someone wants to use `luseepy` with some other Jax code, this may be unexpected. How about we add config variable `enforce_64_bit`? If this flag is False, we don't do anything (so Jax will be 32 bit unless the user set the environment variable); otherwise we keep the current code.

File: lusee/pre_jax/MonoSkyModels.py (hunk 3 above)
Comment (anigmetov) [reply lusee/pre_jax/MonoSkyModels.py:1]: Moving to `tests` sounds better. The idea was to have frozen old code that we use for regression tests. When time comes for major re-write of the Beam class, we can get rid of the pre-Jax Beam.

**Comment by anigmetov** [comment]
@xzackli That seems to be final. I removed 64 bit settings and applied renaming, in the process I realized that the design of SimDriver is not good (inheriting from builtin dict should be replaced with composition), so the last commit ended up being a bit heavier than I anticipated, but it's mostly local changes.

**Review by xzackli — APPROVED** [review]
